### PR TITLE
Add PoolFormer

### PR DESCRIPTION
**Pull request by tanaymeh**
# What does this PR do?

This PR adds the PoolFormer model to the 🤗 repository.

I also opened an Issue for adding the model https://github.com/huggingface/transformers/issues/14584

# Who can review?
@NielsRogge

### CODE REVIEW

File: docs/source/model_doc/poolformer.mdx [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+[[autodoc]] PoolFormerModel
+    - forward
+
+## PoolFormerFeatureExtractor
+[[autodoc]] PoolFormerFeatureExtractor
+    - __call__
+
+[[autodoc]] PoolFormerForImageClassification
+    - forward
```
Comment (NielsRogge) [inline docs/source/model_doc/poolformer.mdx:44]:
```suggestion
## PoolFormerConfig

[[autodoc]] PoolFormerConfig

## PoolFormerFeatureExtractor
[[autodoc]] PoolFormerFeatureExtractor
    - __call__

## PoolFormerModel

[[autodoc]] PoolFormerModel
    - forward

## PoolFormerForImageClassification

[[autodoc]] PoolFormerForImageClassification
    - forward
```

File: src/transformers/__init__.py [hunk 2]
```diff
@@ -150,6 +150,7 @@
         "load_tf2_weights_in_pytorch_model",
     ],
     # Models
+    "models.poolformer": ["POOLFORMER_PRETRAINED_CONFIG_ARCHIVE_MAP", "PoolFormerConfig"],
```
Comment (NielsRogge) [inline src/transformers/__init__.py:153]: Please place this one on the correct alphabetical position

File: src/transformers/__init__.py [hunk 3]
```diff
@@ -638,6 +640,15 @@
...
     # PyTorch models structure
+
+    _import_structure["models.poolformer"].extend(
+        [
+            "POOLFORMER_PRETRAINED_MODEL_ARCHIVE_LIST",
+            "PoolFormerModel",
+            "PoolFormerPreTrainedModel",
+            "PoolFormerForImageClassification",
+        ]
+    )
```
Comment (NielsRogge) [inline src/transformers/__init__.py:651]: Same here.

File: src/transformers/__init__.py [hunk 4]
```diff
@@ -2349,6 +2360,7 @@
         load_tf2_weights_in_pytorch_model,
     )
     from .models.albert import ALBERT_PRETRAINED_CONFIG_ARCHIVE_MAP, AlbertConfig
+    from .models.poolformer import POOLFORMER_PRETRAINED_CONFIG_ARCHIVE_MAP, PoolFormerConfig
```
Comment (NielsRogge) [inline src/transformers/__init__.py:2363]: Same here.

File: src/transformers/__init__.py [hunk 5]
```diff
@@ -2711,6 +2724,12 @@
         from .utils.dummy_pytorch_quantization_and_torch_objects import *
 
     if is_torch_available():
+
+        from .models.poolformer import (
+            POOLFORMER_PRETRAINED_MODEL_ARCHIVE_LIST,
+            PoolFormerModel,
+            PoolFormerPreTrainedModel,
+        )
```
Comment (NielsRogge) [inline src/transformers/__init__.py:2732]: Same here.

File: src/transformers/models/__init__.py [hunk 6]
```diff
@@ -17,6 +17,7 @@
 # limitations under the License.
 
 from . import (
+    poolformer,
```
Comment (NielsRogge) [inline src/transformers/models/__init__.py:20]: Same here.

File: src/transformers/models/auto/modeling_auto.py [hunk 7]
```diff
@@ -159,6 +160,8 @@
 MODEL_WITH_LM_HEAD_MAPPING_NAMES = OrderedDict(
     [
         # Model with LM heads mapping
+
+        ("poolformer", "PoolFormerForConditionalGeneration"),
```
Comment (NielsRogge) [inline src/transformers/models/auto/modeling_auto.py:164]:
```suggestion
```

File: src/transformers/models/auto/modeling_auto.py [hunk 8]
```diff
@@ -213,6 +216,7 @@
 MODEL_FOR_CAUSAL_LM_MAPPING_NAMES = OrderedDict(
     [
         # Model for Causal LM mapping
+        ("poolformer", "PoolFormerForCausalLM"),
```
Comment (NielsRogge) [inline src/transformers/models/auto/modeling_auto.py:219]:
```suggestion
```

File: src/transformers/models/auto/modeling_auto.py [hunk 9]
```diff
@@ -348,6 +352,8 @@
 MODEL_FOR_SEQ_TO_SEQ_CAUSAL_LM_MAPPING_NAMES = OrderedDict(
     [
         # Model for Seq2Seq Causal LM mapping
+
+        ("poolformer", "PoolFormerForConditionalGeneration"),
```
Comment (NielsRogge) [inline src/transformers/models/auto/modeling_auto.py:356]:
```suggestion
```

File: src/transformers/models/auto/modeling_auto.py [hunk 10]
```diff
@@ -376,6 +382,7 @@
 MODEL_FOR_SEQUENCE_CLASSIFICATION_MAPPING_NAMES = OrderedDict(
     [
         # Model for Sequence Classification mapping
+        ("poolformer", "PoolFormerForSequenceClassification"),
```
Comment (NielsRogge) [inline src/transformers/models/auto/modeling_auto.py:385]:
```suggestion
```

File: src/transformers/models/poolformer/configuration_poolformer.py [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+        >>> from transformers import PoolFormerModel, PoolFormerConfig
+
+        >>> # Initializing a PoolFormer sail/poolformer_s12 style configuration
+        >>> configuration = PoolFormerConfig()
+
+        >>> # Initializing a model from the sail/poolformer_s12 style configuration
+        >>> model = PoolFormerModel(configuration)
+
+        >>> # Accessing the model configuration
+        >>> configuration = model.config
```
Comment (NielsRogge) [inline src/transformers/models/poolformer/configuration_poolformer.py:86]:
```suggestion
        
This code example needs to be shifted to the left, otherwise it won't properly render. See configuration_vit.py as an example.

        >>> from transformers import PoolFormerModel, PoolFormerConfig

        >>> # Initializing a PoolFormer sail/poolformer_s12 style configuration
        >>> configuration = PoolFormerConfig()

        >>> # Initializing a model from the sail/poolformer_s12 style configuration
        >>> model = PoolFormerModel(configuration)

        >>> # Accessing the model configuration
        >>> configuration = model.config
```

File: src/transformers/models/poolformer/configuration_poolformer.py [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+            The initializer range for the weights.
+
+    Example:
+
+    ```python
+     >>> from transformers import PoolFormerModel, PoolFormerConfig >>> # Initializing a PoolFormer sail/poolformer_s12
+    style configuration >>> configuration = PoolFormerConfig() >>> # Initializing a model from the sail/poolformer_s12
+    style configuration >>> model = PoolFormerModel(configuration) >>> # Accessing the model configuration >>>
+    configuration = model.config"""
+    model_type = "poolformer"
```
Comment (NielsRogge) [inline src/transformers/models/poolformer/configuration_poolformer.py:81]: `Make style` messed this up locally for me for some reason.

File: src/transformers/models/poolformer/inference.py [hunk 13]
```diff
@@ -0,0 +1,21 @@
...
+image = Image.open(requests.get(url, stream=True).raw)
+
+inputs = feature_extractor(images=image, return_tensors="pt")
+
+outputs = model(**inputs)
+
+predicted_class = outputs.logits.argmax(-1).item()
+print(model.config.id2label[predicted_class])
+
+print(outputs.logits[0, :3])
```
Comment (NielsRogge) [inline src/transformers/models/poolformer/inference.py:21]: This is to be removed, I added this to check out the logits.

File: docs/source/model_doc/poolformer.mdx [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+
+Tips:
+
+<INSERT TIPS ABOUT MODEL HERE>
```
Comment (NielsRogge) [inline docs/source/model_doc/poolformer.mdx:25]: To be filled in. See BERT as an example.

File: src/transformers/models/poolformer/configuration_poolformer.py [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+            The ratio of the number of channels in the output of the MLP to the number of channels in the input.
+        depths (`list`, *optional*, defaults to [2, 2, 6, 2]):
+            The depth of each encoder block.
+        hidden_sizes (`list`, *optional*, defaults to [64, 128, 320, 512]):
+            The hidden sizes of each encoder block.
+        patch_sizes (`list`, *optional*, defaults to [7, 3, 3, 3]):
+            The size of the input patch for each encoder block.
+        strides (`list`, *optional*, defaults to [4, 2, 2, 2]):
+            The stride of the input patch for each encoder block.
+        padding (`list`, *optional*, defaults to [2, 1, 1, 1]):
```
Comment (sgugger) [inline src/transformers/models/poolformer/configuration_poolformer.py:59]: All the lists should be between back-ticks please.

File: src/transformers/models/poolformer/convert_poolformer_timm_to_pytorch.py [hunk 16]
```diff
@@ -0,0 +1,214 @@
+# coding=utf-8
```
Comment (sgugger) [inline src/transformers/models/poolformer/convert_poolformer_timm_to_pytorch.py:1]: Dumb question but why is the file named after timm if its a conversion from the original repo that doesn't look to have any link with timm?

File: docs/source/index.mdx [hunk 17]
```diff
@@ -214,6 +214,7 @@ Flax), PyTorch, and/or TensorFlow.
 |        OpenAI GPT-2         |       ✅       |       ✅       |       ✅        |         ✅         |      ✅      |
 |           Pegasus           |       ✅       |       ✅       |       ✅        |         ✅         |      ✅      |
 |          Perceiver          |       ✅       |       ❌       |       ✅        |         ❌         |      ❌      |
+|         PoolFormer          |       ❌       |       ❌       |       ✅        |         ❌         |      ❌      |
```
Comment (sgugger) [inline docs/source/index.mdx:217]: The model will also need to be put in the README.

File: src/transformers/models/poolformer/modeling_poolformer.py [hunk 18]
```diff
@@ -0,0 +1,498 @@
...
+            blocks.append(nn.ModuleList(layers))
+
+        self.block = nn.ModuleList(blocks)
+
+    def forward(
+        self,
+        pixel_values,
+        output_hidden_states=False,
+        return_dict=True,
+    ):
```
Comment (sgugger) [inline src/transformers/models/poolformer/modeling_poolformer.py:294]:
```suggestion
    def forward(self, pixel_values, output_hidden_states=False, return_dict=True):
```

File: src/transformers/models/poolformer/modeling_poolformer.py [hunk 19]
```diff
@@ -0,0 +1,498 @@
...
+        all_hidden_states = () if output_hidden_states else None
+
+        hidden_states = pixel_values
+        for idx, x in enumerate(zip(self.patch_embeddings, self.block)):
+            embedding_layer, block_layer = x
```
Comment (sgugger) [inline src/transformers/models/poolformer/modeling_poolformer.py:299]: `x` does not seem to be a very well-suited name. `layers`?

File: src/transformers/models/poolformer/modeling_poolformer.py [hunk 20]
```diff
@@ -0,0 +1,498 @@
...
+
+            if output_hidden_states:
+                all_hidden_states = all_hidden_states + (hidden_states,)
+
+        if not return_dict:
+            return tuple(v for v in [hidden_states, all_hidden_states] if v is not None)
+
+        return PoolFormerModelOutput(
+            last_hidden_state=hidden_states,
+            hidden_states=all_hidden_states,
```
Comment (sgugger) [inline src/transformers/models/poolformer/modeling_poolformer.py:315]:
```suggestion
            last_hidden_state=hidden_states, hidden_states=all_hidden_states
```

File: src/transformers/models/poolformer/modeling_poolformer.py [hunk 21]
```diff
@@ -0,0 +1,498 @@
...
+
+        loss = None
+        if labels is not None:
+            if self.num_labels == 1:
+                #  We are doing regression
+                loss_fct = MSELoss()
+                loss = loss_fct(logits.view(-1), labels.view(-1))
+            else:
+                loss_fct = CrossEntropyLoss()
+                loss = loss_fct(logits.view(-1, self.num_labels), labels.view(-1))
```
Comment (sgugger) [inline src/transformers/models/poolformer/modeling_poolformer.py:488]: This should use the config `problem_type`. Please look at the `BertForSequenceClassification` model for an example to follow.

File: src/transformers/models/poolformer/modeling_poolformer.py [hunk 22]
```diff
@@ -0,0 +1,498 @@
...
+
+        if not return_dict:
+            output = (logits,) + outputs[2:]
+            return ((loss,) + output) if loss is not None else output
+
+        return PoolFormerClassifierOutput(
+            loss=loss,
+            logits=logits,
+            hidden_states=outputs.hidden_states,
```
Comment (sgugger) [inline src/transformers/models/poolformer/modeling_poolformer.py:497]:
```suggestion
            loss=loss, logits=logits, hidden_states=outputs.hidden_states
```

File: tests/test_modeling_poolformer.py [hunk 23]
```diff
@@ -0,0 +1,345 @@
...
+class PoolFormerModelTest(ModelTesterMixin, unittest.TestCase):
+
+    all_model_classes = (
+        (
+            PoolFormerModel,
+            PoolFormerForImageClassification,
+        )
+        if is_torch_available()
+        else ()
+    )
```
Comment (sgugger) [inline tests/test_modeling_poolformer.py:134]: This can takes a bit less space ;-)
```suggestion
    all_model_classes = (PoolFormerModel, PoolFormerForImageClassification) if is_torch_available() else ()
```

File: tests/test_modeling_poolformer.py [hunk 24]
```diff
@@ -0,0 +1,345 @@
...
+
+        for model_class in self.all_model_classes:
+            if model_class in get_values(MODEL_MAPPING):
+                continue
+            # TODO: remove the following 3 lines once we have a MODEL_FOR_SEMANTIC_SEGMENTATION_MAPPING
+            # this can then be incorporated into _prepare_for_class in test_modeling_common.py
+            if model_class.__name__ == "PoolFormerForSemanticSegmentation":
+                batch_size, num_channels, height, width = inputs_dict["pixel_values"].shape
+                inputs_dict["labels"] = torch.zeros(
+                    [self.model_tester.batch_size, height, width], device=torch_device
```
Comment (sgugger) [inline tests/test_modeling_poolformer.py:305]: Can we reuse `batch_size` here? Then it could perhaps fit in one line.

**Review by sgugger — APPROVED** [review]
Thanks a lot for working on this!

File: src/transformers/models/poolformer/convert_poolformer_timm_to_pytorch.py (hunk 16 above)
Comment (heytanay) [reply src/transformers/models/poolformer/convert_poolformer_timm_to_pytorch.py:1]: My bad! Will it be right to name it `convert_poolformer_original_to_pytorch.py`?

File: README.md [hunk 25]
```diff
@@ -290,6 +290,7 @@ Min, Patrick Lewis, Ledell Wu, Sergey Edunov, Danqi Chen, and Wen-tau Yih.
...
 1. **[Perceiver IO](https://huggingface.co/docs/transformers/model_doc/perceiver)** (from Deepmind) released with the paper [Perceiver IO: A General Architecture for Structured Inputs & Outputs](https://arxiv.org/abs/2107.14795) by Andrew Jaegle, Sebastian Borgeaud, Jean-Baptiste Alayrac, Carl Doersch, Catalin Ionescu, David Ding, Skanda Koppula, Daniel Zoran, Andrew Brock, Evan Shelhamer, Olivier Hénaff, Matthew M. Botvinick, Andrew Zisserman, Oriol Vinyals, João Carreira.
 1. **[PhoBERT](https://huggingface.co/docs/transformers/model_doc/phobert)** (from VinAI Research) released with the paper [PhoBERT: Pre-trained language models for Vietnamese](https://www.aclweb.org/anthology/2020.findings-emnlp.92/) by Dat Quoc Nguyen and Anh Tuan Nguyen.
+1. **[PoolFormer](https://huggingface.co/docs/transformers/model_doc/poolformer)** (from Sea AI Labs) released with the paper [MetaFormer is Actually What You Need for Vision](https://arxiv.org/abs/2111.11418) by Yu, Weihao and Luo, Mi and Zhou, Pan and Si, Chenyang and Zhou, Yichen and Wang, Xinchao and Feng, Jiashi and Yan, Shuicheng.
```
[the remark this answers is not in the document]
Comment (NielsRogge) [reply README.md:293]: And run `make fix-copies` again :)

File: src/transformers/models/poolformer/convert_poolformer_timm_to_pytorch.py (hunk 16 above)
Comment (NielsRogge) [reply src/transformers/models/poolformer/convert_poolformer_timm_to_pytorch.py:1]: Yes you can rename it like that.

File: tests/test_modeling_poolformer.py [hunk 26]
```diff
@@ -0,0 +1,336 @@
...
+        config.return_dict = True
+
+        for model_class in self.all_model_classes:
+            if model_class in get_values(MODEL_MAPPING):
+                continue
+            # TODO: remove the following 3 lines once we have a MODEL_FOR_SEMANTIC_SEGMENTATION_MAPPING
+            # this can then be incorporated into _prepare_for_class in test_modeling_common.py
+            if model_class.__name__ == "PoolFormerForSemanticSegmentation":
+                batch_size, num_channels, height, width = inputs_dict["pixel_values"].shape
+                inputs_dict["labels"] = torch.zeros([batch_size, height, width], device=torch_device).long()
```
Comment (NielsRogge) [inline tests/test_modeling_poolformer.py:297]: This should be removed.

**Review by LysandreJik — APPROVED** [review]
This looks good to me, thank you for working on it @heytanay!

**Comment by NielsRogge** [comment]
Thanks for all your work on this, merging!
